### DISCUSSION TITLE
Why is an update downloading so many files?

### DISCUSSION THREAD

**Discussion by AArnott** [Squirrel/Squirrel.Windows#1766] [Q&A]
My Squirrel-installed app was at 0.2.35 and was upgrading to 0.2.93. But in the process it appears to have downloaded every single interim version, delta files and full packages. Why would it do that? I could imagine it downloading all the delta packages or just the last full package. The upgrade took quite a long time, BTW, and I wonder if it's because besides downloading it was perhaps also expanding all these packages.

See screenshot.

![image](https://user-images.githubusercontent.com/3548/141605409-46a13619-0dc7-4104-a327-6c19cb2c16db.png)

**Comment by anaisbetts** [comment]
So, I'm not sure what _happened_, but how it works is:

1. We download every delta update from A => B
2. We try to rebuild full package B, by stepping through each delta update building full packages along the way
3. If rebuilding full package B fails, we download full package B

I suspect because it was _so_ much work to get from A => B, Step 2 probably was canceled or failed along the way, and we fell back to step 3

That being said, I would argue that you have perhaps overautomated your release process, as you will grind a **lot** of your users CPU if you upgrade this way (delta updates are written optimized for reliability / making very very sure we get The File We Expect, which means we end up doing a lot of extra copying). I would consider only generating an update stream for explicit releases rather than making a release for every merge to master

**Reply by AArnott** [reply]
On step 3, shouldn't it leap to to the final version if it's to download the entire package? As it was, it downloaded *several* full packages as you can see. 

Regarding my over-automation, it's simply a matter of channels. I have a release channel which updates infrequently. But I also have a CI channel that I personally subscribe to for dogfooding, verifying updates work, etc. After all, I'd hate to learn that updates would fail only _after_ I issue a release. And the case above isn't unique to CI releases. If I had released several times over the course of a year and someone was way behind and finally updated, they presumably still have several releases to leap for the update and would have this negative experience, which is why I dogfood (so I can catch this and resolve it.)
